src/extensions/default/NavigationAndHistory/NavigationProvider.js

```diff
@@ -0,0 +1,553 @@
+            }
+        } else {
+            fileEntry.exists(function (err, exists) {
+                if (!err && exists) {
+                    deferred.resolve();
+                } else {
+                    deferred.reject();
+                }
+            });
+        }
+
+        return deferred.promise();
+    }
+        
+   /**
+    * Prototype to capture a navigation frame and it's various data/functional attributues
+    */
+    function NavigationFrame(editor, selectionObj) {
+        this.cm = editor._codeMirror;
+        this.file = editor.document.file._path;
+        this.inMem = editor.document.file.constructor.name === "InMemoryFile";
+        this.paneId = editor._paneId;
+        this.uId = (new Date()).getTime();
+        this.selections = [];
+        this.bookMarkIds = [];
+        this._createMarkers(selectionObj.ranges);
+    }
+    
+   /**
+    * Lifecycle event handler of the editor for which this frame is captured
+    */
+    NavigationFrame.prototype._handleEditorDestroy = function (editor) {
+        this._backupSelectionRanges();
+        this._clearMarkers();
+        this.cm = null;
+        this.bookMarkIds = null;
+    };
+    
+    /**
+    * Function to re-create CM TextMarkers for previously backed up ranges
+    * This logic is required to ensure that the captured navigation positions 
+    * stay valid and contextual even when the actual document text mutates.
+    * The mutations which are handled here :
+    * -> Addition/Deletion of lines before the captured position
+    * -> Addition/Updation of characters in the captured selection
+    */
+    NavigationFrame.prototype._reinstateMarkers = function (editor) {
+        this.cm = editor._codeMirror;
+        this.paneId = editor._paneId;
+        
+        var range,
+            index,
+            bookMark;
+
+        this.bookMarkIds = [];
+        for (index in this.selections) {
+            range = this.selections[index];
+            // 'markText' has to used for a non-zero length position, if current selection is 
+            // of zero length use bookmark instead.
+            if (range.start.line === range.end.line && range.start.ch === range.end.ch) {
+                bookMark = this.cm.setBookmark(range.start, range.end);
+                this.bookMarkIds.push(bookMark.id);
+            } else {
+                this.cm.markText(range.start, range.end, {className: (this.uId)});
+            }
+        }
+    };
+    
+   /**
+    * Function to create CM TextMarkers for the navigated positions/selections.
+    * This logic is required to ensure that the captured navigation positions 
+    * stay valid and contextual even when the actual document text mutates.
+    * The mutations which are handled here :
+    * -> Addition/Deletion of lines before the captured position
+    * -> Addition/Updation of characters in the captured selection
+    */
+    NavigationFrame.prototype._createMarkers = function (ranges) {
```

Comment:
createMarkers and reinstateMarkers share a lot of code. Maybe we can consolidate them a little more?

As I see it, this should work:
``` js
NavigationFrame.prototype._reinstateMarkers = function (editor) {
    this.cm = editor._codeMirror;
    this.paneId = editor._paneId;
    this._createMarkers(this.selections)
}
```